sql/src/main/java/io/crate/executor/transport/distributed/DownstreamOperationContext.java

```diff
@@ -67,46 +79,59 @@ public void onFailure(Throwable t) {
         });
         this.streamers = streamers;
         this.doneCallback = doneCallback;
+        BucketPage bucketPage = new BucketPage(upstreamFutures);
+        downstreamOperation.pageDownstream().nextPage(bucketPage, new PageConsumeListener() {
+            // PageConsumeCallback
+            @Override
+            public void finish() {
+                needsMoreRows.set(false);
+                finishContext();
+            }
+
+            @Override
+            public void needMore() {
+                // in the current state, we only have one page
+                // so we can finish here
+                finishContext();
+            }
+        });
     }
 
     public void addFailure(@Nullable Throwable failure) {
+        int bucketNum = mergeOperationsLeft.decrementAndGet();
+        LOGGER.trace("bucket #{} failed", bucketNum);
+
         if (failure != null) {
-            logger.error("addFailure local", failure);
+            LOGGER.error("addFailure local", failure);
         } else {
             failure = new UnknownUpstreamFailure();
         }
-        try {
-            boolean firstFailure = listener.setException(failure);
-            logger.trace("addFailure first: {}", firstFailure);
-        } finally {
-            if (mergeOperationsLeft.decrementAndGet() == 0) {
-                doneCallback.finished();
-                downstreamOperation.finished();
-            }
+        if (bucketNum > 0 && !alreadyFinished.get()) {
+            upstreamFutures.get(bucketNum).setException(failure);
         }
+        // TODO: needed?
+        boolean firstFailure = listener.setException(failure);
+        LOGGER.trace("addFailure first: {}", firstFailure);
     }
 
-    public void add(Bucket rows) {
-        assert rows != null;
-        logger.trace("add rows.size: {}", rows.size());
-        synchronized (lock) {
-            if (needsMoreRows) {
-                try {
-                    needsMoreRows = downstreamOperation.addRows(rows);
-                } catch (Exception e) {
-                    logger.error("failed to add rows to downstreamOperation", e);
-                    listener.setException(e);
-                }
-            }
-        }
-
-        if (mergeOperationsLeft.decrementAndGet() == 0) {
-            doneCallback.finished();
-            downstreamOperation.finished();
+    public void add(Bucket bucket) {
+        assert bucket != null;
+        int bucketNum = mergeOperationsLeft.decrementAndGet();
+        LOGGER.trace("add bucket #{} size: {}", bucketNum, bucket.size());
```

Comment:
if we have a concise bucket toString it would make sense to just use bucket as arg in the log
maybe also better for all other cases, since no computation is done if nothing gets logged